common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java

```diff
@@ -368,4 +385,59 @@ private static void closeQuietly(Closeable c) {
+
+    private static final class NoexecFileStoreDetector {
+        static final Method MOUNT_ENTRY_METHOD;
+        static final Method HAS_OPTION_METHOD;
+
+        static {
+            Method mountEntryMethod = null;
+            Method hasOptionMethod = null;
+            if (PlatformDependent.javaVersion() >= 7) {
+                try {
+                    mountEntryMethod = Class.forName("sun.nio.fs.UnixFileStore").getDeclaredMethod("entry");
+                    hasOptionMethod =
+                            Class.forName("sun.nio.fs.UnixMountEntry").getDeclaredMethod("hasOption", String.class);
+
+                    Throwable mountEntryThrowable = ReflectionUtil.trySetAccessible(mountEntryMethod);
+                    Throwable hasOptionThrowable = ReflectionUtil.trySetAccessible(hasOptionMethod);
+                    if (mountEntryThrowable != null || hasOptionThrowable != null) {
+                        // Module accessibility in Java9 will cause a java.lang.reflect.InaccessibleObjectException,
+                        // not a security exception, which directly extends java.lang.RuntimeException, but
+                        // ReflectionUtil.trySetAccessible returns both instead of throwing them.
+                        if (!(mountEntryThrowable instanceof SecurityException) ||
+                            !(hasOptionThrowable instanceof SecurityException)) {
+                            logger.debug("Failed to initialize NativeLibraryLoader.NoexecFileStoreDetector; " +
+                                         "use --add-opens java.base/sun.nio.fs=... to get noexec mount warnings");
+                        }
+                        mountEntryMethod = null;
+                        hasOptionMethod = null;
+                        // Re-throw the security exception to log it
+                        throw mountEntryMethod == null? hasOptionThrowable : mountEntryThrowable;
+                    }
+                } catch (ClassNotFoundException e) {
+                    // Normal on non-Unix operating systems or on Java < 7
+                } catch (Throwable t) {
+                    logger.debug("Failed to initialize NativeLibraryLoader.NoexecFileStoreDetector", t);
+                }
+            }
+            MOUNT_ENTRY_METHOD = mountEntryMethod;
+            HAS_OPTION_METHOD = hasOptionMethod;
+        }
+
+        @SuppressForbidden(reason = "to obtain mount information about file stores")
+        private static boolean isOnNoexecFileStore(File file)
+                throws IOException, InvocationTargetException, IllegalAccessException {
+            if (PlatformDependent.javaVersion() < 7 || HAS_OPTION_METHOD == null || MOUNT_ENTRY_METHOD == null) {
+                return false;
+            }
+            FileStore fileStore = Files.getFileStore(file.toPath());
+            Object unixMountEntry = MOUNT_ENTRY_METHOD.invoke(fileStore);
+            return (Boolean) HAS_OPTION_METHOD.invoke(unixMountEntry, "noexec");
```

Comment:
I guess there is no way to detect an issue by avoiding internal APIs?
- Checking if the directory is readable and/or [File#canExecute](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#canExecute())
- Use [FileStore#getAttribute](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileStore.html#getAttribute(java.lang.String)) or some other attribute API offered by NIO